File: src/sc_neurocore/neurons/models/siegert.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Siegert 1951 — mean-field LIF firing rate

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Any

import numpy as np


@dataclass
class SiegertTransferFunction:
    """Siegert 1951 — mean-field LIF firing rate.

    Analytical stationary firing rate of a LIF neuron driven by
    Gaussian white noise: r = [tau_rp + tau_m * sqrt(pi) *
    integral(exp(u^2)*(1+erf(u)), u_reset..u_thresh)]^{-1}
    Uses Gauss-Hermite quadrature approximation.

    Reference: Siegert, A.J.F. (1951). Phys. Rev. 81:617–623.
    """

    tau_m: float = 20.0  # ms, membrane time constant
    tau_rp: float = 2.0  # ms, refractory period
    v_threshold: float = -50.0  # mV
    v_reset: float = -70.0  # mV
    v_rest: float = -65.0  # mV

    def __post_init__(self) -> None:
        for field in ("v_threshold", "v_reset", "v_rest"):
            value = getattr(self, field)
            if not math.isfinite(value):
                raise ValueError(f"{field} must be finite")
        for field in ("tau_m", "tau_rp"):
            value = getattr(self, field)
            if not math.isfinite(value) or value <= 0.0:
                raise ValueError(f"{field} must be positive and finite")
        if self.v_threshold <= self.v_reset:
            raise ValueError("v_threshold must be greater than v_reset")

    def step(self, current: float) -> float:
        """Return instantaneous firing rate (Hz) for given mean input current."""
        if not math.isfinite(current):
            raise ValueError("current must be finite")
        self._validate_runtime_state()
        mu = self.v_rest + current
        if not math.isfinite(mu):
            raise ValueError("Siegert runtime mean voltage must remain finite")
        sigma = max(abs(current) * 0.1, 1e-6)
        if not math.isfinite(sigma) or sigma <= 0.0:
            raise ValueError("Siegert runtime diffusion scale must remain finite and positive")
        u_th = (self.v_threshold - mu) / sigma
        u_re = (self.v_reset - mu) / sigma
        if not math.isfinite(u_th) or not math.isfinite(u_re) or u_th <= u_re:
            raise ValueError("Siegert runtime first-passage bounds must be finite and ordered")
        # Gauss-Legendre quadrature over [u_re, u_th]
        n_quad = 40
        u_pts, w_pts = np.polynomial.legendre.leggauss(n_quad)
        half_range = 0.5 * (u_th - u_re)
        mid = 0.5 * (u_th + u_re)
        if not math.isfinite(half_range) or not math.isfinite(mid) or half_range <= 0.0:
            raise ValueError("Siegert runtime quadrature interval must remain finite")
        u_scaled = half_range * u_pts + mid
        integrand = np.exp(np.clip(u_scaled**2, None, 50.0)) * (1.0 + _erf_approx(u_scaled))
        if not np.all(np.isfinite(integrand)):
            raise ValueError("Siegert runtime integrand must remain finite")
        integral_val = float(half_range * np.sum(w_pts * integrand))
        if not math.isfinite(integral_val) or integral_val < 0.0:
            raise ValueError("Siegert runtime integral must remain finite and non-negative")
        t_isi = self.tau_rp + self.tau_m * np.sqrt(np.pi) * integral_val
        if not math.isfinite(t_isi) or t_isi < self.tau_rp:
            raise ValueError("Siegert runtime inter-spike interval must remain finite")
        rate = 1000.0 / t_isi
        max_rate = 1000.0 / self.tau_rp
        if not math.isfinite(rate) or rate < 0.0 or rate > max_rate:
            raise ValueError("Siegert runtime rate must remain finite and refractory-bounded")
        return float(rate)  # Hz

    def reset(self) -> None:
        pass

    def _validate_runtime_state(self) -> None:
        try:
            self.__post_init__()
        except ValueError as exc:
            raise ValueError(f"Siegert runtime parameters invalid: {exc}") from exc


def _erf_approx(x: np.ndarray[Any, Any]) -> np.ndarray[Any, Any]:
    """Abramowitz & Stegun 7.1.26 rational approximation."""
    sign = np.sign(x)
    a = np.abs(x)
    p = 0.3275911
    t = 1.0 / (1.0 + p * a)
    coeffs = np.array([0.254829592, -0.284496736, 1.421413741, -1.453152027, 1.061405429])
    poly = t * (coeffs[0] + t * (coeffs[1] + t * (coeffs[2] + t * (coeffs[3] + t * coeffs[4]))))
    result: np.ndarray[Any, Any] = sign * (1.0 - poly * np.exp(-a * a))
    return result

File: src/sc_neurocore/accel/julia/neurons/siegert.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia for siegert

module SiegertAccel

export step!, simulate, validate_siegert, SiegertTransferFunctionState

mutable struct SiegertTransferFunctionState
    tau_m::Float64
    tau_rp::Float64
    v_threshold::Float64
    v_reset::Float64
    v_rest::Float64
end

function SiegertTransferFunctionState()
    SiegertTransferFunctionState(20.0, 2.0, -50.0, -70.0, -65.0)
end

function step!(s::SiegertTransferFunctionState, I_ext::Float64=0.0; dt::Float64=0.1)
    if !validate_siegert(s) || !isfinite(I_ext)
        throw(DomainError((s.tau_m, s.tau_rp, I_ext), "Siegert state/current must be finite and physically ordered"))
    end

    mu = s.v_rest + I_ext
    if !isfinite(mu)
        throw(DomainError(mu, "Siegert mean voltage must remain finite"))
    end
    sigma = max(abs(I_ext) * 0.1, 1e-06)
    if !isfinite(sigma) || sigma <= 0.0
        throw(DomainError(sigma, "Siegert diffusion scale must remain finite and positive"))
    end
    u_th = (s.v_threshold - mu) / sigma
    u_re = (s.v_reset - mu) / sigma
    if !isfinite(u_th) || !isfinite(u_re) || u_th <= u_re
        throw(DomainError((u_re, u_th), "Siegert first-passage bounds must remain finite and ordered"))
    end
    n_quad = 40
    (u_pts, w_pts) = legendre_nodes_weights(n_quad)
    half_range = 0.5 * (u_th - u_re)
    mid = 0.5 * (u_th + u_re)
    if !isfinite(half_range) || !isfinite(mid) || half_range <= 0.0
        throw(DomainError((half_range, mid), "Siegert quadrature interval must remain finite"))
    end
    integral_val = 0.0
    for i in eachindex(u_pts)
        u_scaled = half_range * u_pts[i] + mid
        integrand = exp(clamp(u_scaled ^ 2, -Inf, 50.0)) * (1.0 + erf_approx(u_scaled))
        if !isfinite(integrand)
            throw(DomainError(integrand, "Siegert integrand must remain finite"))
        end
        integral_val += w_pts[i] * integrand
    end
    integral_val *= half_range
    if !isfinite(integral_val) || integral_val < 0.0
        throw(DomainError(integral_val, "Siegert integral must remain finite and non-negative"))
    end
    t_isi = s.tau_rp + s.tau_m * sqrt(pi) * integral_val
    if !isfinite(t_isi) || t_isi < s.tau_rp
        throw(DomainError(t_isi, "Siegert inter-spike interval must remain finite and refractory bounded"))
    end
    rate = 1000.0 / t_isi
    max_rate = 1000.0 / s.tau_rp
    if !isfinite(rate) || rate < 0.0 || rate > max_rate
        throw(DomainError(rate, "Siegert rate must remain finite and refractory bounded"))
    end
    return rate
end

function validate_siegert(s::SiegertTransferFunctionState)
    return isfinite(s.tau_m) && s.tau_m > 0.0 && isfinite(s.tau_rp) && s.tau_rp > 0.0 &&
           isfinite(s.v_threshold) && isfinite(s.v_reset) && isfinite(s.v_rest) &&
           s.v_threshold > s.v_reset
end

function erf_approx(x::Float64)
    sign_x = sign(x)
    a = abs(x)
    p = 0.3275911
    t = 1.0 / (1.0 + p * a)
    poly = t * (0.254829592 + t * (-0.284496736 + t * (1.421413741 + t * (-1.453152027 + t * 1.061405429))))
    return sign_x * (1.0 - poly * exp(-a * a))
end

function legendre_nodes_weights(n::Int)
    if n != 40
        error("only 40-point quadrature is supported")
    end
    nodes = [-0.9982377097105593, -0.990726238699457, -0.9772599499837743, -0.9579168192137917, -0.9328128082786765, -0.9020988069688743, -0.8659595032122595, -0.8246122308333117, -0.7783056514265194, -0.7273182551899271, -0.6719566846141796, -0.6125538896679802, -0.5494671250951282, -0.4830758016861787, -0.413779204371605, -0.3419940908257585, -0.2681521850072537, -0.1926975807013711, -0.1160840706752552, -0.03877241750605082, 0.03877241750605082, 0.1160840706752552, 0.1926975807013711, 0.2681521850072537, 0.3419940908257585, 0.413779204371605, 0.4830758016861787, 0.5494671250951282, 0.6125538896679802, 0.6719566846141796, 0.7273182551899271, 0.7783056514265194, 0.8246122308333117, 0.8659595032122595, 0.9020988069688743, 0.9328128082786765, 0.9579168192137917, 0.9772599499837743, 0.990726238699457, 0.9982377097105593]
    weights = [0.004521277098533191, 0.010498284531152813, 0.01642105838190789, 0.022245849194166957, 0.027937006980023402, 0.03346019528254785, 0.03878216797447202, 0.04387090818567327, 0.04869580763507223, 0.053227846983936824, 0.05743976909939155, 0.06130624249292894, 0.06480401345660104, 0.0679120458152339, 0.07061164739128678, 0.07288658239580406, 0.07472316905796826, 0.07611036190062624, 0.07703981816424797, 0.07750594797842481, 0.07750594797842481, 0.07703981816424797, 0.07611036190062624, 0.07472316905796826, 0.07288658239580406, 0.07061164739128678, 0.0679120458152339, 0.06480401345660104, 0.06130624249292894, 0.05743976909939155, 0.053227846983936824, 0.04869580763507223, 0.04387090818567327, 0.03878216797447202, 0.03346019528254785, 0.027937006980023402, 0.022245849194166957, 0.01642105838190789, 0.010498284531152813, 0.004521277098533191]
    return nodes, weights
end

function simulate(n_steps::Int=1000; I_ext::Float64=10.0, dt::Float64=0.1)
    s = SiegertTransferFunctionState()
    trace = zeros(n_steps)
    spikes = 0
    for t in 1:n_steps
        result = step!(s, I_ext; dt=dt)
        trace[t] = result
        if result isa Number && result > 0
            spikes += 1
        end
    end
    return trace, spikes
end

end # module SiegertAccel
